### PR TITLE
Added support for methods that return reference (Issue #139)

### PR DESCRIPTION
**Pull request by edsonmedina**

### CODE REVIEW

**Comment by aik099** [comment]
Closes #139

File: src/Prophecy/Doubler/Generator/Node/MethodNode.php [hunk 1]
```diff
@@ -71,6 +72,16 @@ public function setStatic($static = true)
         $this->static = (bool) $static;
     }
 
+    public function returnsReference()
+    {
+        return $this->returnsReference;
+    }
+
+    public function setReturnsReference($value = true)
```
Comment (stof) [inline src/Prophecy/Doubler/Generator/Node/MethodNode.php]: the argument should not be called ``$value``

**Comment by stof** [comment]
I see several issues here:

- this should be covered by specs (and btw, the current specs are failing because of the extra calls you added)
- I'm quite sure that this does not fix the issue entirely, because the method code will not be compatible with returning by reference AFAICT
- in any case, mocking a function returning by ref will not give you a reference to anything else than a local variable inside the double. doing something else would require Prophecy to return by reference everywhere (which would create weird issues for all other cases)

**Comment by edsonmedina** [comment]
* Specs fixed.
* How does it currently deal with arguments passed by reference?

**Comment by stof** [comment]
@edsonmedina they are not reference anymore inside the Prophecy layer, so you cannot change it by reference

**Comment by edsonmedina** [comment]
I agree that this does not solve the issue completely, but at least it doesn't break anymore when mocking a class/interface with a reference returning method. Currently it dies even if you're not touching that specific method.

My patch won't allow you to return references, but it will allow you to extend the class/interface and mock other stuff. Which is better than nothing IMO.

**Comment by stof** [comment]
But it will still break if you call the method

**Comment by stof** [comment]
the ClassMirrorSpec and the ClassGeneratorSpec are still missing specs about the behavior of functions returning by references (if you revert your ClassGenerator change, the spec will still pass)

**Comment by edsonmedina** [comment]
The reason why i'm fixing this is I'm trying to mock an external dependency on a project. The dependency has a method (which is not called at all) that returns by reference. Just by extending the class I was getting a fatal error (even if I'm never going to use that method). 

Sure, if I call the method now, weird things will happen.

So what do you suggest? Maybe throwing an exception when calling the method?

(I'll add specs for the new methods' behaviour)

**Comment by everzet** [comment]
@edsonmedina throwing an exception sounds fine by me.

**Comment by edsonmedina** [comment]
I think I screwed up with a rebase, I'll do another clean push request, sorry.

**Comment by everzet** [comment]
Okay :)

**Pull request closed without merging**